### DISCUSSION TITLE
Has anyone tried to add compreFace vectors to milvus?

### DISCUSSION THREAD

**Discussion by martinenkoEduard** [milvus-io/milvus#18734] [Q&A and General discussion] [answered]
What metrics does compreface use to compare face similarity?
Does it use Euclidean distance (L2)? Or something different?

How can I improve search results using milvus?
Is it better to use
Euclidean distance (L2)
or
Inner product (IP)
?

Maybe somehow change vectors(normalize or something else) before inserting to milvus?

What distance metric algorithm compreface uses to compare two faces?
Does it somehow changes vectors?

Maybe someone has such experience?

**Comment by martinenkoEduard** [comment]
Do I need to use 
faiss.normalize_L2(normalized)

**Comment by yhmo** [comment, marked as answer]
- Milvus supports two types of metric types for float vectors: L2(Euclidean distance) and IP(Inner product).
- The IP metric type requires all the vectors to be normalized before being inserted into Milvus.
- If vectors are normalized, you will get the same result no matter IP or L2 metric type.
- Not sure which library you mean for the "compareface". I think the L2 and IP is the most common way to calculate distance between two vectors, and most of these kinds of libraries ought to support them.

**Reply by martinenkoEduard** [reply]
compreface  https://github.com/exadel-inc/CompreFace
Face recognition software.

Btw doesn't normalized vectors with IP metric equal to Cosine Similarity?

**Reply by yhmo** [reply]
> compreface https://github.com/exadel-inc/CompreFace Face recognition software.
> 
> Btw doesn't normalized vectors with IP metric equal to Cosine Similarity?

![Screenshot from 2022-08-23 10-01-56](https://user-images.githubusercontent.com/2282099/186052372-866fc07a-34da-42ac-82d4-551a27778d2a.png)

If u and v are normalized, |u|=1, |v|=1, then Cosine(u, v)=u.v. So, yes, normalized vectors with IP metric equal to cosine similarity.
